bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingReadOp.java

```diff
@@ -110,6 +110,7 @@
 
         @Override
         public void close() {
+            complete.set(true);
```

Comment:
wouldn't it be safer to use `if (complete.compareAndSet(false, true)) {` to shield against races?

Reply:
addressed by https://github.com/apache/bookkeeper/pull/3110#issuecomment-1068857090

Reply:
Now that we have the `writeSet.recycle();` block above we could go back to compareAndSet()

What about this ?

```
if (complete.compareAndSet(false, true)) {
   writeSet.recycle();
}
```

This way:
- we do writeSet.recycle(); only if complete is "false"
- eventually complete will be "true" in any case

Reply:
```
            if (!complete.get()) {
                writeSet.recycle();
            }
            complete.set(true);
```
I think that would also work

Reply:
@congbobo184 with this approach we risk calling recycle twice if close() is called from different threads. 
At least we need to implement @eolivelli 's suggestion. I'd set rc to something as well, eg:

```
if (complete.compareAndSet(false, true)) {
   rc = BKException.Code.UnexpectedConditionException;
   writeSet.recycle();
}
```

Reply:
@congbobo184 this is the last comment, then we are good to go (and also cut a release)